compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs

```diff
@@ -604,13 +595,69 @@ instance Pretty Context where
+      , _relatedInformation = Nothing
+      }
+
+data Warning
+  = WContext !Context !Warning
+  | WTemplateChangedPrecondition !TypeConName
+  | WTemplateChangedSignatories !TypeConName
+  | WTemplateChangedObservers !TypeConName
+  | WTemplateChangedAgreement !TypeConName
+  | WChoiceChangedControllers !ChoiceName
+  | WChoiceChangedObservers !ChoiceName
+  | WChoiceChangedAuthorizers !ChoiceName
+  | WTemplateChangedKeyExpression !TypeConName
+  | WTemplateChangedKeyMaintainers !TypeConName
+  | WTemplateAddedKeyDefinition !TypeConName !TemplateKey
+  | WCouldNotExtractForUpgradeChecking !T.Text !(Maybe T.Text)
+    -- ^ When upgrading, we extract relevant expressions for things like
+    -- signatories. If the expression changes shape so that we can't get the
+    -- underlying expression that has changed, this warning is emitted.
+  deriving (Show)
+
+warningLocation :: Warning -> Maybe SourceLoc
+warningLocation = \case
+  WContext ctx _ -> contextLocation ctx
+  _ -> Nothing
+
+instance Pretty Warning where
+  pPrint = \case
+    WContext ctx err ->
+      vcat
+      [ "warning while type checking " <> pretty ctx <> ":"
+      , nest 2 (pretty err)
+      ]
+    WTemplateChangedPrecondition template -> "The upgraded template " <> pPrint template <> " cannot change the definition of its precondition."
+    WTemplateChangedSignatories template -> "The upgraded template " <> pPrint template <> " cannot change the definition of its signatories."
+    WTemplateChangedObservers template -> "The upgraded template " <> pPrint template <> " cannot change the definition of its observers."
+    WTemplateChangedAgreement template -> "The upgraded template " <> pPrint template <> " cannot change the definition of agreement."
+    WChoiceChangedControllers choice -> "The upgraded choice " <> pPrint choice <> " cannot change the definition of controllers."
+    WChoiceChangedObservers choice -> "The upgraded choice " <> pPrint choice <> " cannot change the definition of observers."
+    WChoiceChangedAuthorizers choice -> "The upgraded choice " <> pPrint choice <> " cannot change the definition of authorizers."
+    WTemplateChangedKeyExpression template -> "The upgraded template " <> pPrint template <> " cannot change the expression for computing its key."
+    WTemplateChangedKeyMaintainers template -> "The upgraded template " <> pPrint template <> " cannot change the maintainers for its key."
+    WTemplateAddedKeyDefinition template _key -> "The upgraded template " <> pPrint template <> " cannot add a key if it didn't have one previously."
+    WCouldNotExtractForUpgradeChecking attribute mbExtra -> "Could not check if the upgrade of " <> text attribute <> " is valid because its expression is the wrong shape." <> foldMap (const " Extra context: " <> text) mbExtra
+
+instance ToDiagnostic Warning where
+  toDiagnostic warning = Diagnostic
+      { _range = maybe noRange sourceLocToRange (warningLocation warning)
+      , _severity = Just DsWarning
```

Comment:
later on it might be a good idea to add a flag (or even detect `--ghc-option=-Werror`) to upgrade warnings to errors

Reply:
Is this the only place in the compiler where we emit warnings? If not, do we detect `--ghc-option=-Werror` in these other cases?